### PR TITLE
Feature: Better Bookmark Searching

### PR DESCRIPTION
**Pull request by JoshAshby**
# TL;DR

transientBug's search interface currently sucks and has a lot of room for improvement.

# Info
I've been looking for a way to build a better search interface into transientBug which takes advantage of the various information that transientBug stores, such as the tags, title, and description, along with parsed URI and offline-cached content. I've tried a few different approaches including implicitly searching all fields behind the scenes (what currently happens, resulting in pretty abysmal results), building a text query parser for queries like `title:"something in the title" -tags:this_is_a_tag +created_at:>2014-01-01` which got rather messy because parsing user-supplied text is never clean, to finally building this interface which exposes all the same query refinement power of the text-based one, while avoiding having to parse user-supplied input.

Behind the scenes, this implements a React based component that renders out a form that POST's back to the server (code in `app/javascript/query_builder`). This form builds a simplified AST which is transformed through a "Searcher" class (`app/searchers/application_searcher.rb` and `app/searchers/elasticsearch_searcher.rb`) into an ES query, which finally runs through Chewy against the bookmarks index to generate the search results. To save the query and be able to re-render it on a reload, this creates a `Search` record and redirects the browser to the new record, which isn't ideal but also the only way I could think of to get around the limitation of getting the query to the server and back when it can't go over a GET request in the query string parameters.

Unfortunately, this does make the front end more dependant on javascript, but it also provides me with some good grounds for doing more progressive enhancement of the UI through smaller components rather than full-fledged SPA's because reasons.

# TODO

- [ ] Finish Tests
- [ ] Document Searcher and frontend code
- [X] Figure out how to deploy this since public/packs is ignored. I could unignore it, or figure out a build process that does the API docs and asset compilation for me

# Pretty Image
transientbug searching bookmarks 2018-06-19 14-47-07

### CODE REVIEW

**Comment by JoshAshby** [comment]
Current hot mess (but working) of code demo:
screencast 2018-06-20 13-40-18

**Comment by JoshAshby** [comment]
I'm thinking of breaking this, and the other pr #39, into three PR's where:
* One is the searcher interface, providing an interface for converting a generic query AST into ES or SQL/ActiveModel queries
* One is the graphical interface presented in this PR, with progressive enhancement of the site (without javascript the search is just always against title, description, and tags)
* One introduces the textual parser from #39 which converts it into both a better graphical interface AND the generic AST 
That way every ones happy and I get to write ALL THE CODE

But I'm not sure. 🤷‍♀️

File: app/controllers/bookmarks/search_controller.rb [hunk 1]
```diff
@@ -1,19 +1,85 @@
...
 
     respond_to do |format|
       format.html { render :index }
       format.json { render :index, status: :ok }
     end
   end
+
+  # POST /bookmarks/search
+  # POST /bookmarks/search.json
+  # rubocop:disable Metrics/AbcSize
```
Comment (robacarp) [inline app/controllers/bookmarks/search_controller.rb]: ಠ_ಠ

File: app/controllers/bookmarks/search_controller.rb [hunk 2]
```diff
@@ -1,19 +1,85 @@
...
+
+  protected
+
+  def set_search
+    @search = authorize current_user.searches.find(params[:id])
+  end
+
+  # rubocop:disable Naming/MemoizedInstanceVariableName
+  def default_query
+    return @query ||= {} if params[:q].blank? || params[:q].empty?
```
Comment (robacarp) [inline app/controllers/bookmarks/search_controller.rb]: this logic is weird

File: app/javascript/query_builder/components/Clause.jsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        <Form.Select
+          options={ operationOptions }
+          onChange={ this.selectOperation }
+          value={ this.props.query.operation }
+        />
+
+        <input type="hidden" name={ `${ this.props.root }[operation]` } value={ this.props.query.operation } />
+
+        { this.currentValues.map((value, i) => (
+          /* eslint react/no-array-index-key:0 */
```
Comment (robacarp) [inline app/javascript/query_builder/components/Clause.jsx]: heh

File: app/policies/search_policy.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+class SearchPolicy < ApplicationPolicy
+  class Scope < Scope
+    def resolve
+      # return scope.where(public: true) unless user.present?
```
Comment (robacarp) [inline app/policies/search_policy.rb]: ❌

File: app/searchers/application_searcher/concerns/active_record.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+class ApplicationSearcher
+  module Concerns
+    module ActiveRecord
+      extend ActiveSupport::Concern
+
+      included do
+      end
```
Comment (robacarp) [inline app/searchers/application_searcher/concerns/active_record.rb]: ?

File: config/initializers/content_security_policy.rb [hunk 6]
```diff
@@ -1,12 +1,16 @@
 Rails.application.config.content_security_policy do |p|
-  p.default_src :self, :https
+  p.default_src :self, :https, "http://localhost:*", "ws://localhost:*", "http://localhost:8080"
```
Comment (robacarp) [inline config/initializers/content_security_policy.rb]: doooooooo you want these in here?

File: app/controllers/bookmarks/search_controller.rb (hunk 1 above)
Comment (JoshAshby) [reply app/controllers/bookmarks/search_controller.rb]: I should learn to work with these :|

File: config/initializers/content_security_policy.rb (hunk 6 above)
Comment (JoshAshby) [reply config/initializers/content_security_policy.rb]: Thought I had made these dev only, good catch

File: app/controllers/bookmarks/search_controller.rb (hunk 2 above)
Comment (JoshAshby) [reply app/controllers/bookmarks/search_controller.rb]: This feature is weird. Its a fallback for when no js is enabled but I should see about encapsulating it into the searcher class, maybe?
